# Exercise 4.7

# 01 importing Libraries

In [4]:
# importing pandas, numpy and os
import pandas as pd
import numpy as np
import os

# 02 User-Defined Functions

In [6]:
# Def function Defines add_num as a function with two variables x and y
def add_num(x,y):
#add_num given definition as sum and the variables x and y to be added together
    sum = x + y
# I want the result of the sum to be printed(returned)
    return sum 

In [7]:
# Running Function
add_num(9,11)

20

# 03 Importing Dataframe

## creating path

In [10]:
# path creation
path = r'/Users/Andreas/Desktop/13-02-2025 Instacart Basket Analysis/02 Data'

In [11]:
# testing path
path

'/Users/Andreas/Desktop/13-02-2025 Instacart Basket Analysis/02 Data'

## importing orders_products_merged.pkl

In [13]:
# importing dataframe
df_o_p_m = pd.read_pickle(os.path.join(path, 'Prepared Data', 'orders_products_merged.pkl'))

In [14]:
# testing dataframe
df_o_p_m.head()

product_id                product_name  aisle_id  department_id    prices  \
0           1  Chocolate Sandwich Cookies        61             19  5.800781   
1           1  Chocolate Sandwich Cookies        61             19  5.800781   
2           1  Chocolate Sandwich Cookies        61             19  5.800781   
3           1  Chocolate Sandwich Cookies        61             19  5.800781   
4           1  Chocolate Sandwich Cookies        61             19  5.800781   

   order_id  user_id  order_number  orders_day_of_week  order_hour_of_day  \
0   3139998      138            28                   6                 11   
1   1977647      138            30                   6                 17   
2    389851      709             2                   0                 21   
3    652770      764             1                   3                 13   
4   1813452      764             3                   4                 17   

   days_without_ordering  add_to_cart_order  reordered _merge  \
0                    3.0                  5          0   both   
1                   20.0                  1          1   both   
2                    6.0                 20          0   both   
3                    0.0                 10          0   both   
4                    9.0                 11          1   both   

         price_label     busiest_day         busiest_days  
0  Mid-range product  Regularly busy  Regularly busy days  
1  Mid-range product  Regularly busy  Regularly busy days  
2  Mid-range product     Busiest day         Busiest days  
3  Mid-range product  Regularly busy      Least busy days  
4  Mid-range product      Least busy      Least busy days

In [15]:
df_o_p_m.info()

<class 'pandas.core.frame.DataFrame'>
RangeIndex: 32404859 entries, 0 to 32404858
Data columns (total 17 columns):
 #   Column                 Dtype   
---  ------                 -----   
 0   product_id             int32   
 1   product_name           object  
 2   aisle_id               int16   
 3   department_id          int16   
 4   prices                 float16 
 5   order_id               int32   
 6   user_id                int32   
 7   order_number           int16   
 8   orders_day_of_week     int16   
 9   order_hour_of_day      int16   
 10  days_without_ordering  float16 
 11  add_to_cart_order      int16   
 12  reordered              int16   
 13  _merge                 category
 14  price_label            object  
 15  busiest_day            object  
 16  busiest_days           object  
dtypes: category(1), float16(2), int16(7), int32(3), object(4)
memory usage: 1.9+ GB


In [16]:
#changing data types to accomadate lack of RAM
df_o_p_m['product_id']= df_o_p_m['product_id'].astype('int32')
df_o_p_m['aisle_id'] = df_o_p_m['aisle_id'].astype('int16')
df_o_p_m['department_id'] = df_o_p_m['department_id'].astype('int16')
df_o_p_m['prices'] = df_o_p_m['prices'].astype('float64')
df_o_p_m['order_id'] = df_o_p_m['order_id'].astype('int32')
df_o_p_m['user_id'] = df_o_p_m['user_id'].astype('int32')
df_o_p_m['order_number'] = df_o_p_m['order_number'].astype('int16')
df_o_p_m['orders_day_of_week'] = df_o_p_m['orders_day_of_week'].astype('int16')
df_o_p_m['order_hour_of_day'] = df_o_p_m['order_hour_of_day'].astype('int16')
df_o_p_m['days_without_ordering'] = df_o_p_m['days_without_ordering'].astype('float16')
df_o_p_m['add_to_cart_order'] = df_o_p_m ['add_to_cart_order'].astype('int16')
df_o_p_m['reordered'] = df_o_p_m['reordered'].astype('int16')

In [17]:
# shape of dataframe
df_o_p_m.info()

<class 'pandas.core.frame.DataFrame'>
RangeIndex: 32404859 entries, 0 to 32404858
Data columns (total 17 columns):
 #   Column                 Dtype   
---  ------                 -----   
 0   product_id             int32   
 1   product_name           object  
 2   aisle_id               int16   
 3   department_id          int16   
 4   prices                 float64 
 5   order_id               int32   
 6   user_id                int32   
 7   order_number           int16   
 8   orders_day_of_week     int16   
 9   order_hour_of_day      int16   
 10  days_without_ordering  float16 
 11  add_to_cart_order      int16   
 12  reordered              int16   
 13  _merge                 category
 14  price_label            object  
 15  busiest_day            object  
 16  busiest_days           object  
dtypes: category(1), float16(1), float64(1), int16(7), int32(3), object(4)
memory usage: 2.1+ GB


In [18]:
df_o_p_m.shape

(32404859, 17)

# 04 Creating Subset of Data for User-defined function practice

In [20]:
# Creating a Subset of the first 1,000,000 rows of df_o_p_m
df = df_o_p_m[:1000000]
# if : placed at the end it would give you 1,000,000 rows starting from bottom of table

In [21]:
# Checking output
df.shape

(1000000, 17)

# 05 User Defined Function to Group Data

In [23]:
# Creating function price_label to analyze every row (row)
def price_label(row):

#if value is < or = to 5 return as Low-range product
  if row['prices'] <= 5:
    return 'Low-range product'
#else if value is > than 5 and < or = 15 to return Mid-range product
  elif (row['prices'] > 5) and (row['prices'] <= 15):
    return 'Mid-range product'
#else if value is > than 15 return High-range product
  elif row['prices'] > 15:
    return 'High-range product'
# if no value found in row return Not Enough Data
  else: return 'Not enough data'

In [24]:
# using price_label in df on the ['price_range'] column starting at the first row (1)
df['price_range'] = df.apply(price_label, axis = 1)

/var/folders/bz/xnfhbvns50zc0x2wgk6fnbg80000gq/T/ipykernel_63815/3004831912.py:2: SettingWithCopyWarning: 
A value is trying to be set on a copy of a slice from a DataFrame.
Try using .loc[row_indexer,col_indexer] = value instead

See the caveats in the documentation: https://pandas.pydata.org/pandas-docs/stable/user_guide/indexing.html#returning-a-view-versus-a-copy
  df['price_range'] = df.apply(price_label, axis = 1)


In [25]:
# checking output
df['price_range'].value_counts(dropna = False) 

price_range
Mid-range product     652638
Low-range product     338018
High-range product      9344
Name: count, dtype: int64

In [26]:
# checking shape (should have 1 extra column = price_range)
df.shape

(1000000, 18)

In [27]:
# checking the max price to make sure its above 15
df['prices'].max()

24.5

# 06 User-Defined Function using loc()

In [29]:
# creating price_range_loc column with its identifier high, medium or low-range product

#perform loc() on df / the prices in df are being compared to > 15 IF that is true THEN = High-range product
df_o_p_m.loc[df_o_p_m['prices'] > 15, 'price_label'] = 'High-range product'

#perform loc() on df / the prices in df are being compared to <= 15 or < 5 IF that is true THEN = Mid-range product
df_o_p_m.loc[(df_o_p_m['prices'] <= 15) & (df_o_p_m['prices'] > 5), 'price_label'] = 'Mid-range product' 

#perform loc() on df / the prices in df are being compared to <= 5 IF that is true THEN = Low-range product
df_o_p_m.loc[df_o_p_m['prices'] <= 5, 'price_label'] = 'Low-range product'

#all the results are place in the 'price_range_loc' column created

In [30]:
# output check
df_o_p_m['price_label'].value_counts(dropna = False)

price_label
Mid-range product     21860860
Low-range product     10126321
High-range product      417678
Name: count, dtype: int64

# 07 For Loops

In [32]:
# first I need to find which day is the busiest for orders
df_o_p_m['orders_day_of_week'].value_counts(dropna=False)

orders_day_of_week
0    6204182
1    5660230
6    4496490
2    4213830
5    4205791
3    3840534
4    3783802
Name: count, dtype: int64

In [33]:
# Creating an if-loop for asigning busiest, least, regularly busy 

#first step to create empty shell for the loop to return value to
result = []

#specifying for a value in df_o_p_m on column orders day of week IF value = 0 THEN return Busiest Day
for value in df_o_p_m["orders_day_of_week"]:
  if value == 0:
    result.append("Busiest day")
  
#ELSE IF value in df_o_p_m = 4 THEN return least busy
  elif value == 4:
    result.append("Least busy")
 
#else IF non of the above then return regularly busy
  else:
    result.append("Regularly busy")

# This is much faster than user-defined function which searches all of a dataframe.

In [34]:
# now we need to add the list we have created (result) onto df_o_p_m

#place the results into a column named busiest_day
df_o_p_m['busiest_day'] = result

In [35]:
#checking output
df_o_p_m.head()

product_id                product_name  aisle_id  department_id    prices  \
0           1  Chocolate Sandwich Cookies        61             19  5.800781   
1           1  Chocolate Sandwich Cookies        61             19  5.800781   
2           1  Chocolate Sandwich Cookies        61             19  5.800781   
3           1  Chocolate Sandwich Cookies        61             19  5.800781   
4           1  Chocolate Sandwich Cookies        61             19  5.800781   

   order_id  user_id  order_number  orders_day_of_week  order_hour_of_day  \
0   3139998      138            28                   6                 11   
1   1977647      138            30                   6                 17   
2    389851      709             2                   0                 21   
3    652770      764             1                   3                 13   
4   1813452      764             3                   4                 17   

   days_without_ordering  add_to_cart_order  reordered _merge  \
0                    3.0                  5          0   both   
1                   20.0                  1          1   both   
2                    6.0                 20          0   both   
3                    0.0                 10          0   both   
4                    9.0                 11          1   both   

         price_label     busiest_day         busiest_days  
0  Mid-range product  Regularly busy  Regularly busy days  
1  Mid-range product  Regularly busy  Regularly busy days  
2  Mid-range product     Busiest day         Busiest days  
3  Mid-range product  Regularly busy      Least busy days  
4  Mid-range product      Least busy      Least busy days

In [36]:
# checking output
df_o_p_m['busiest_day'].value_counts(dropna=False)

busiest_day
Regularly busy    22416875
Busiest day        6204182
Least busy         3783802
Name: count, dtype: int64

In [37]:
# checking that all values are same
df_o_p_m['busiest_day'].value_counts(dropna=False).sum()

32404859

In [38]:
# checking all values same as above
df_o_p_m['price_label'].value_counts(dropna = False).sum()

32404859

# 08 TASK

## Question 2

    suppose your clients have changed their minds about the labels you created in your “busiest_day” column. Now, they want “Busiest day” to become “Busiest days” (plural). This label should correspond with the two busiest days of the week as opposed to the single busiest day. At the same time, they’d also like to know the two slowest days. Create a new column for this using a suitable method.

In [41]:
result = []

#specifying for a value in df_o_p_m on column orders day of week IF value = 0 or 1 THEN return Busiest Day
for value in df_o_p_m["orders_day_of_week"]:
  if value == 0 or value == 1:
    result.append("Busiest days")
  
#ELSE IF value in df_o_p_m = 4 or 3 THEN return least busy
  elif value == 4 or value == 3:
    result.append("Least busy days")
 
#else IF non of the above then return regularly busy
  else:
    result.append("Regularly busy days")

In [42]:
#placing the results of our loop into newly created busiest days column
df_o_p_m['busiest_days'] = result

In [43]:
#checking output
df_o_p_m['busiest_days'].value_counts(dropna=False)

busiest_days
Regularly busy days    12916111
Busiest days           11864412
Least busy days         7624336
Name: count, dtype: int64

In [44]:
#checking frequency
df_o_p_m['busiest_days'].value_counts(dropna=False).sum()

32404859


## Answer

busiest_day

    Regularly busy    22416875
    Busiest day        6204182
    Least busy         3783802
    Name: count, dtype: int64


busiest_days

    Regularly busy days    12916111
    Busiest days           11864412
    Least busy days         7624336
    Name: count, dtype: int64



Having grouped more days together the number of variables classed as busiest and least busy days have increased whilst the number of variables classed as regularly busy has decreased

## Question 4 

    When too many users make Instacart orders at the same time, the app freezes. The senior technical officer at Instacart wants you to identify the busiest hours of the day. Rather than by hour, they want periods of time labeled “Most orders,” “Average orders,” and “Fewest orders.” Create a new column containing these labels called “busiest_period_of_day.”

In [47]:
#first we need to find the frequency of the order_hour_of_day
df_o_p_m['order_hour_of_day'].value_counts(dropna = False)

order_hour_of_day
10    2761760
11    2736140
14    2689136
15    2662144
13    2660954
12    2618532
16    2535202
9     2454203
17    2087654
8     1718118
18    1636502
19    1258305
20     976156
7      891054
21     795637
22     634225
23     402316
6      290493
0      218769
1      115700
5       87961
2       69375
4       53242
3       51281
Name: count, dtype: int64

In [48]:
# now we want to create a result list to hold the variables we create through loop

result = []

#we want the loop to run in the order_hour_of_day column in df_o_p_m

for value in df_o_p_m['order_hour_of_day']:

#IF value is >= 9 or <= 16 THEN = Most Orders   
    if value >= 9 and value <=16:
        result.append('Most Orders')

#ELSE IF a value is (7 or 8) or (17 to 23) THEN = average orders    
    elif (value >=7 and value <=8) or (value >=17 and value <=23):
        result.append('Average Orders')

# If non of the above apply THEN = most orders    
    else:
        result.append('Fewest Orders')

In [49]:
#creating a column for created list (result)
df_o_p_m['busiest_period_of_day'] = result

In [50]:
df_o_p_m['busiest_period_of_day'].value_counts(dropna=False)

busiest_period_of_day
Most Orders       21118071
Average Orders    10399967
Fewest Orders       886821
Name: count, dtype: int64

# Export

In [52]:
df_o_p_m.to_pickle(os.path.join(path, 'Prepared Data', 'orders_products_merged2.pkl'))